File: backend/schemas/insumos/Costos.py
// src/services/insumosCostosService.ts
export type CostoInsumo = {
  id: string;
  fecha: string;
  categoria: "Medicamento/Vacuna" | "Alimento" | "Equipo/Herramienta" | "Limpieza" | "Suministro General";
  presupuesto: number;
  real: number;
  variacion: number;
  comentarios: string;
};

const STORAGE_KEY = "insumos_costos";

function getAll(): CostoInsumo[] {
  const raw = localStorage.getItem(STORAGE_KEY);
  return raw ? JSON.parse(raw) : [];
}

function saveAll(lista: CostoInsumo[]) {
  localStorage.setItem(STORAGE_KEY, JSON.stringify(lista));
}

export function getCostos() {
  return getAll();
}

export function addCosto(costo: Omit<CostoInsumo, "id"|"variacion">) {
  const variacion = costo.real - costo.presupuesto;
  const nuevo: CostoInsumo = { ...costo, id: crypto.randomUUID(), variacion };
  const costos = getAll();
  costos.push(nuevo);
  saveAll(costos);
  return nuevo;
}

export function updateCosto(id: string, costo: Omit<CostoInsumo, "id"|"variacion">) {
  const costos = getAll();
  const idx = costos.findIndex(c => c.id === id);
  if (idx >= 0) {
    const variacion = costo.real - costo.presupuesto;
    costos[idx] = { ...costo, id, variacion };
    saveAll(costos);
    return costos[idx];
  }
  throw new Error("Registro no encontrado");
}

export function deleteCosto(id: string) {
  const costos = getAll().filter(c => c.id !== id);
  saveAll(costos);
}

File: frontend/src/services/insumos/Costos.ts
// src/services/Costos.ts
import API_BASE from "../../config/api";

export type CostoInsumo = {
  id: number;
  fecha: string;
  modulo: string;
  categoria:
    | "Medicamento/Vacuna"
    | "Alimento"
    | "Equipo/Herramienta"
    | "Limpieza"
    | "Suministro General";
  insumo: string;
  lote?: string | null;
  cantidad: number;
  unidad: string;
  costoUnitario: number;
  costoTotal: number;
  proveedor?: string | null;
  descripcion: string;
  presupuesto?: number;
  real?: number;
  variacion?: number;
};

const EMPRESA_ID = 1;
const GRANJA_ID = 1;

function mapApiToCosto(api: any): CostoInsumo {
  const costoTotal = api.costo_total ?? 0;
  const presupuesto = api.presupuesto ?? costoTotal;
  const real = api.real ?? costoTotal;
  const variacion = real - presupuesto;

  return {
    id: api.id,
    fecha: api.fecha,
    modulo: api.modulo,
    categoria: api.categoria,
    insumo: api.insumo,
    lote: api.lote ?? null,
    cantidad: api.cantidad,
    unidad: api.unidad,
    costoUnitario: api.costo_unitario,
    costoTotal,
    proveedor: api.proveedor ?? null,
    descripcion: api.descripcion ?? "",
    presupuesto,
    real,
    variacion,
  };
}

export async function getCostos(): Promise<CostoInsumo[]> {
  const res = await fetch(
    `${API_BASE}/insumos/costos/?empresa_id=${EMPRESA_ID}&granja_id=${GRANJA_ID}`
  );
  if (!res.ok) {
    throw new Error("Error al obtener costos de insumos");
  }
  const data = await res.json();
  return data.map(mapApiToCosto);
}

export type NuevoCostoInsumo = {
  fecha: string;
  modulo: string;
  categoria: CostoInsumo["categoria"];
  insumo: string;
  lote?: string;
  cantidad: number;
  unidad: string;
  costoUnitario: number;
  costoTotal: number;
  proveedor?: string;
  descripcion: string;
};

export async function addCosto(
  costo: NuevoCostoInsumo
): Promise<CostoInsumo> {
  const payload = {
    empresa_id: EMPRESA_ID,
    granja_id: GRANJA_ID,
    fecha: costo.fecha,
    modulo: costo.modulo,
    categoria: costo.categoria,
    insumo: costo.insumo,
    lote: costo.lote || null,
    cantidad: costo.cantidad,
    unidad: costo.unidad,
    costo_unitario: costo.costoUnitario,
    costo_total: costo.costoTotal,
    proveedor: costo.proveedor || null,
    descripcion: costo.descripcion,
  };

  const res = await fetch(`${API_BASE}/insumos/costos/`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  if (!res.ok) {
    throw new Error("Error al registrar costo de insumo");
  }
  const data = await res.json();
  return mapApiToCosto(data);
}

export async function updateCosto(
  id: number,
  costo: NuevoCostoInsumo
): Promise<CostoInsumo> {
  const payload = {
    fecha: costo.fecha,
    modulo: costo.modulo,
    categoria: costo.categoria,
    insumo: costo.insumo,
    lote: costo.lote || null,
    cantidad: costo.cantidad,
    unidad: costo.unidad,
    costo_unitario: costo.costoUnitario,
    costo_total: costo.costoTotal,
    proveedor: costo.proveedor || null,
    descripcion: costo.descripcion,
  };

  const res = await fetch(
    `${API_BASE}/insumos/costos/${id}?empresa_id=${EMPRESA_ID}&granja_id=${GRANJA_ID}`,
    {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    }
  );
  if (!res.ok) {
    throw new Error("Error al actualizar costo de insumo");
  }
  const data = await res.json();
  return mapApiToCosto(data);
}

export async function deleteCosto(id: number): Promise<void> {
  const res = await fetch(
    `${API_BASE}/insumos/costos/${id}?empresa_id=${EMPRESA_ID}&granja_id=${GRANJA_ID}`,
    { method: "DELETE" }
  );
  if (!res.ok) {
    throw new Error("Error al eliminar costo de insumo");
  }
}
